sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go

```diff
@@ -0,0 +1,127 @@
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package expansionx contains utilities for starting expansion services for
+// cross-language transforms.
+package expansionx
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+)
+
+type url string
+
+type jarGetter struct {
+	repository url
+	groupID    string
+	jarCache   string
+}
+
+var defaultJarGetter = newJarGetter()
+
+const (
+	apacheRepository = url("https://repo.maven.apache.org/maven2")
+	beamGroupID      = "org/apache/beam"
+	jarCache         = "~/.apache_beam/cache/jars"
+)
+
+func newJarGetter() *jarGetter {
+	return &jarGetter{repository: apacheRepository, groupID: beamGroupID, jarCache: jarCache}
+}
+
+// GetBeamJar checks a temporary directory for the desired Beam JAR, downloads the
+// appropriate JAR from Maven if not present, then returns the file path to the
+// JAR.
+func GetBeamJar(gradleTarget, version string) (string, error) {
+	return defaultJarGetter.getJar(gradleTarget, version)
+}
+
+func (j *jarGetter) getJar(gradleTarget, version string) (string, error) {
+	strippedTarget := dropEndOfGradleTarget(gradleTarget)
+	fullURL, jarName := j.getURLForBeamJar(strippedTarget, version)
+
+	cacheDir := j.getCacheDir()
+	err := os.MkdirAll(cacheDir, 0700)
+	if err != nil {
+		return "", err
+	}
+
+	jarPath := filepath.Join(cacheDir, jarName)
+
+	if jarExists(jarPath) {
+		return jarPath, nil
+	}
+
+	resp, err := http.Get(string(fullURL))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("Received non 200 response code, got %v", resp.StatusCode)
+	}
+
+	file, err := os.Create(jarPath)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return jarPath, nil
+}
+
+// getURLForBeamJar builds the Maven URL for the JAR and the JAR name, returning both
+// separately so the JAR name can be used for saving the file later.
+func (j *jarGetter) getURLForBeamJar(gradleTarget, version string) (url, string) {
+	baseURL := j.repository + url("/"+j.groupID+"/")
+	fullTarget := "beam" + gradleTarget
+	targetPath := strings.ReplaceAll(fullTarget, ":", "-")
+	jarName := strings.ReplaceAll(fullTarget, ":", "-") + "-" + version + ".jar"
```

Comment:
I see what you mean, there's a bit of ordering here that can be improved, and so many "

I think in this instance, fmt.Sprintf is probably marginally more efficient than the concat too. But we don't care too much about efficiency here, readability first, efficiency later.

Since you don't use baseURL outside of the final URL, there's no need to do it early. And we can use the `path` package which always uses forward slashes for the joins.  https://pkg.go.dev/path#Join

On the whole it's probably cleaner to do this:

```
gradlePath := strings.ReplaceAll(gradleTarget, ":", "-")
targetPath := "beam" + gradlePath
jarName :=  fmt.Sprintf("beam%s-%s.jar", targetPath,version)
finalURL := url(path.Join(j.repository, j.groupID, targetPath, version, jarName))
```

Only 4 pairs of quotes too.